libraries/CryptoInterface/src/CryptoInterface.cpp

```diff
@@ -0,0 +1,566 @@
+  {
+    assert(ctMaxDataLength() - ctMinDataLength <= CT_MAX_DIFF);
+    _ctMinDataLength = ctMinDataLength; 
+  }
+  size_t ctMinDataLength() {return _ctMinDataLength;}
+
+  void setCtMaxDataLength(const size_t ctMaxDataLength) 
+  { 
+    assert(ctMaxDataLength - ctMinDataLength() <= CT_MAX_DIFF);
+    _ctMaxDataLength = ctMaxDataLength; 
+  }
+  size_t ctMaxDataLength() {return _ctMaxDataLength;}
+
+  void setWarningsEnabled(bool warningsEnabled) { _warningsEnabled = warningsEnabled; }
+  bool warningsEnabled() { return _warningsEnabled; }
+
+  void setNonceGenerator(nonceGeneratorType nonceGenerator) { _nonceGenerator = nonceGenerator; }
+  nonceGeneratorType getNonceGenerator() { return _nonceGenerator; }
+
+
+  // #################### MD5 ####################
+    
+  // resultArray must have size MD5_NATURAL_LENGTH or greater
+  void *md5Hash(const void *data, const size_t dataLength, void *resultArray)
+  {
+    if(warningsEnabled())
+      Serial.println(F("\nWARNING! The MD5 hash is broken in terms of attacker resistance.\n" 
+                       "Only use it in those cases where attacker resistance is not important. Prefer SHA-256 or higher otherwise.\n"
+                       "Use CryptoInterface::setWarningsEnabled(false) to turn off this warning.\n"));
+    
+    br_md5_context context;
+    br_md5_init(&context);
+    br_md5_update(&context, data, dataLength);
+    br_md5_out(&context, resultArray);
+    return resultArray;
+  }
+  
+  String md5Hash(const String &message)
+  {
+    uint8_t hash[MD5_NATURAL_LENGTH];
+    md5Hash(message.c_str(), message.length(), hash);
+    return TypeCast::uint8ArrayToHexString(hash, MD5_NATURAL_LENGTH);
+  }
+
+  void *md5Hmac(const void *data, const size_t dataLength, const void *hashKey, const size_t hashKeyLength, void *resultArray, const size_t outputLength)
+  {  
+    return createBearsslHmac(&br_md5_vtable, data, dataLength, hashKey, hashKeyLength, resultArray, outputLength);
+  }
+  
+  String md5Hmac(const String &message, const void *hashKey, const size_t hashKeyLength, const size_t hmacLength)
+  {
+    return createBearsslHmac(&br_md5_vtable, MD5_NATURAL_LENGTH, message, hashKey, hashKeyLength, hmacLength);
+  }
+
+  void *md5HmacCT(const void *data, const size_t dataLength, const void *hashKey, const size_t hashKeyLength, void *resultArray, const size_t outputLength)
+  {
+    return createBearsslHmacCT(&br_md5_vtable, data, dataLength, hashKey, hashKeyLength, resultArray, outputLength);
+  }
+  
+  String md5HmacCT(const String &message, const void *hashKey, const size_t hashKeyLength, const size_t hmacLength)
+  {
+    return createBearsslHmacCT(&br_md5_vtable, MD5_NATURAL_LENGTH, message, hashKey, hashKeyLength, hmacLength);
+  }
+
+
+  // #################### SHA-1 ####################
+
+  // resultArray must have size SHA1_NATURAL_LENGTH or greater
+  void *sha1Hash(const void *data, const size_t dataLength, void *resultArray)
+  {
+    if(warningsEnabled())
```

Comment:
This should probably have a `deprecated` applied to all the sha1 functions in this case.  Get compiler warnings no matter what.

The warningsEnabled() really needs to be a compile-time constexpr.  Suggest a DEBUG-type macro instead.  Don't want these strings in code for prod ships...

Reply:
Excellent idea. I added the deprecated attribute to the MD5 and SHA1 hash functions, and moved the old warnings to the comments.

HMAC seems to be safe even though the underlying hash function is broken in terms of collision resistance [[1]](https://crypto.stackexchange.com/questions/9336/is-hmac-md5-considered-secure-for-authenticating-encrypted-data)[[2]](https://tools.ietf.org/html/rfc6151)[[3]](https://stackoverflow.com/questions/2889473/when-is-it-safe-to-use-a-broken-hash-function). So the warning is only applicable to the hash functions for now, although it is naturally prudent to choose a less broken underlying hash function if possible, just in case.